Source/WebCore/page/Navigation.cpp

```diff
@@ -147,6 +147,49 @@ void Navigation::updateForActivation(HistoryItem* previousItem, std::optional<Na
     m_activation = NavigationActivation::create(*type, *currentEntry(), WTFMove(previousEntry));
 }
 
+// https://html.spec.whatwg.org/multipage/browsing-the-web.html#fire-the-pageswap-event
+RefPtr<NavigationActivation> Navigation::createForPageswapEvent(HistoryItem* newItem, DocumentLoader* documentLoader)
+{
+    auto type = documentLoader->triggeringAction().navigationAPIType();
+    if (!type)
+        return nullptr;
+
+    bool isSameOrigin = SecurityOrigin::create(documentLoader->documentURL())->isSameOriginAs(window()->protectedDocument()->securityOrigin());
+    if (!isSameOrigin || !documentLoader->request().isSameSite())
```

Comment:
I would add a comment here that you are checking `.isSameSite()` specifically to verify no cross-origin redirect has happened.